File: shared/go/bridge_client.go
// Package stitch provides shared primitives for Stitch Go clients.
//
// All Go bridge clients (go-python, go-ruby, go-nodejs) import this package
// instead of duplicating scanner, pending-map, kill, and ready-wait logic.
package stitch

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os/exec"
	"sync"
	"syscall"
	"time"
)

// ─────────────────────────────────────────────────────────────────────────────
// Protocol types
// ─────────────────────────────────────────────────────────────────────────────

// RpcError carries the error payload from the sidecar.
type RpcError struct {
	Message   string `json:"message"`
	Traceback string `json:"traceback,omitempty"`
	Backtrace string `json:"backtrace,omitempty"`
	Code      int    `json:"code,omitempty"`
}

func (e *RpcError) Error() string {
	if e.Traceback != "" {
		return fmt.Sprintf("%s\n%s", e.Message, e.Traceback)
	}
	if e.Backtrace != "" {
		return fmt.Sprintf("%s\n%s", e.Message, e.Backtrace)
	}
	return e.Message
}

// RpcResponse is the wire format for incoming JSON-RPC responses.
type RpcResponse struct {
	ID     string          `json:"id"`
	Result json.RawMessage `json:"result,omitempty"`
	Error  *RpcError       `json:"error,omitempty"`
}

// RpcChunk is emitted by the sidecar for streaming responses.
// Chunk frames carry data; the terminal frame has Result set (empty object = stream done).
type RpcChunk struct {
	ID    string          `json:"id"`
	Chunk json.RawMessage `json:"chunk,omitempty"`
	// Result is set on the terminal frame (empty object = stream done).
	Result json.RawMessage `json:"result,omitempty"`
	Error  *RpcError       `json:"error,omitempty"`
}

// ─────────────────────────────────────────────────────────────────────────────
// IsChunk - detect streaming chunk frames
// ─────────────────────────────────────────────────────────────────────────────

// IsChunk reports whether the raw JSON is a streaming chunk frame
// (has "chunk" key but no "result" or "error" key).
// When true, callers should dispatch to StreamPendingMap instead of PendingMap.
func IsChunk(raw []byte) bool {
	var m map[string]json.RawMessage
	if err := json.Unmarshal(raw, &m); err != nil {
		return false
	}
	_, hasChunk := m["chunk"]
	_, hasResult := m["result"]
	_, hasError := m["error"]
	return hasChunk && !hasResult && !hasError
}

// ─────────────────────────────────────────────────────────────────────────────
// PendingMap - mutex-guarded map of in-flight calls
// ─────────────────────────────────────────────────────────────────────────────

// PendingMap manages in-flight RPC calls, each identified by a string UUID.
// It is safe for concurrent use from multiple goroutines.
// Note: PendingMap.Dispatch handles RpcResponse frames (result or error).
// For streaming chunk frames (detected via IsChunk), dispatch to StreamPendingMap instead.
type PendingMap struct {
	mu    sync.Mutex
	calls map[string]chan RpcResponse
}

// NewPendingMap creates an initialised PendingMap.
func NewPendingMap() *PendingMap {
	return &PendingMap{calls: make(map[string]chan RpcResponse)}
}

// Register inserts a channel for the given request id and returns it.
// The returned channel is buffered with capacity 1.
func (p *PendingMap) Register(id string) chan RpcResponse {
	ch := make(chan RpcResponse, 1)
	p.mu.Lock()
	p.calls[id] = ch
	p.mu.Unlock()
	return ch
}

// Dispatch delivers resp to the channel registered for resp.ID.
// It is a no-op if no channel is registered (e.g. caller timed out).
func (p *PendingMap) Dispatch(resp RpcResponse) {
	p.mu.Lock()
	ch, ok := p.calls[resp.ID]
	if ok {
		delete(p.calls, resp.ID)
	}
	p.mu.Unlock()
	if ok {
		ch <- resp
	}
}

// Delete removes the entry for id without delivering a response.
// Use this when a call is cancelled or timed out before a reply arrives.
func (p *PendingMap) Delete(id string) {
	p.mu.Lock()
	delete(p.calls, id)
	p.mu.Unlock()
}

// DrainWithError unblocks all pending callers with an error response.
// Call this when the child process exits unexpectedly.
func (p *PendingMap) DrainWithError(message string) {
	p.mu.Lock()
	for id, ch := range p.calls {
		ch <- RpcResponse{ID: id, Error: &RpcError{Message: message}}
		delete(p.calls, id)
	}
	p.mu.Unlock()
}

// ─────────────────────────────────────────────────────────────────────────────
// StreamPendingMap - in-flight streaming RPC calls
// ─────────────────────────────────────────────────────────────────────────────

// StreamPendingMap manages in-flight streaming RPC calls.
// Each call is identified by a string UUID and receives frames on a buffered channel.
type StreamPendingMap struct {
	mu    sync.Mutex
	calls map[string]chan RpcChunk
}

// NewStreamPendingMap creates an initialised StreamPendingMap.
func NewStreamPendingMap() *StreamPendingMap {
	return &StreamPendingMap{calls: make(map[string]chan RpcChunk)}
}

// Register creates a buffered channel for streaming id. The channel receives
// one RpcChunk per chunk frame plus a terminal frame (with Result set).
func (s *StreamPendingMap) Register(id string) chan RpcChunk {
	ch := make(chan RpcChunk, 32) // buffer to avoid blocking the reader
	s.mu.Lock()
	s.calls[id] = ch
	s.mu.Unlock()
	return ch
}

// Dispatch delivers frame to the channel registered for frame.ID.
// If the frame is terminal (Result or Error set), the entry is removed.
// It is a no-op if no channel is registered for that ID.
func (s *StreamPendingMap) Dispatch(frame RpcChunk) {
	s.mu.Lock()
	ch, ok := s.calls[frame.ID]
	if ok && (frame.Result != nil || frame.Error != nil) {
		// terminal frame - deliver and remove
		delete(s.calls, frame.ID)
	}
	s.mu.Unlock()
	if ok {
		ch <- frame
	}
}

// Delete removes the entry for id without delivering a frame.
func (s *StreamPendingMap) Delete(id string) {
	s.mu.Lock()
	delete(s.calls, id)
	s.mu.Unlock()
}

// ─────────────────────────────────────────────────────────────────────────────
// Stream - streaming JSON-RPC call
// ─────────────────────────────────────────────────────────────────────────────

// Stream sends a streaming JSON-RPC request and returns a channel that
// receives chunk frames. The channel is closed after the terminal frame.
// The caller is responsible for consuming or draining the returned channel.
//
// Usage:
//
//	ch, err := Stream(streamPending, id, writeFunc)
//	for frame := range ch {
//	    if frame.Error != nil { /* handle error */ }
//	    // frame.Chunk has the data; frame.Result != nil means done
//	}
func Stream(
	streamPending *StreamPendingMap,
	id string,
	writeFunc func() error,
) (<-chan RpcChunk, error) {
	ch := streamPending.Register(id)
	if err := writeFunc(); err != nil {
		streamPending.Delete(id)
		return nil, err
	}

	// Return a wrapper channel that closes itself when the terminal frame arrives.
	out := make(chan RpcChunk, 32)
	go func() {
		defer close(out)
		for frame := range ch {
			out <- frame
			if frame.Result != nil || frame.Error != nil {
				return // terminal
			}
		}
	}()
	return out, nil
}

// ─────────────────────────────────────────────────────────────────────────────
// CallOptions - per-call behaviour
// ─────────────────────────────────────────────────────────────────────────────

// CallOptions control per-call behaviour.
type CallOptions struct {
	// Context for cancellation and deadlines.
	Context context.Context
	// Traceparent is the W3C trace context header value.
	// If non-empty it is included as "traceparent" in the request JSON.
	Traceparent string
}

// WithTraceparent returns a CallOptions with the given W3C traceparent value.
func WithTraceparent(t string) CallOptions {
	return CallOptions{Traceparent: t, Context: context.Background()}
}

// ─────────────────────────────────────────────────────────────────────────────
// Scanner constructor with enlarged buffer
// ─────────────────────────────────────────────────────────────────────────────

const defaultScannerBufSize = 4 * 1024 * 1024 // 4 MiB

// NewScanner returns a *bufio.Scanner reading from r with a 4 MiB buffer.
// This avoids scanner.ErrTooLong on large JSON payloads.
func NewScanner(r io.Reader) *bufio.Scanner {
	s := bufio.NewScanner(r)
	s.Buffer(make([]byte, defaultScannerBufSize), defaultScannerBufSize)
	return s
}

// ─────────────────────────────────────────────────────────────────────────────
// KillChild - SIGTERM → SIGKILL(2 s) pattern
// ─────────────────────────────────────────────────────────────────────────────

// KillChild sends SIGTERM to the child process; if it has not exited after 2 s
// it sends SIGKILL.  Safe to call when cmd.Process is nil.
func KillChild(cmd *exec.Cmd) {
	if cmd.Process == nil {
		return
	}

	_ = cmd.Process.Signal(syscall.SIGTERM)

	done := make(chan struct{})
	go func() {
		_ = cmd.Wait()
		close(done)
	}()

	timer := time.NewTimer(2 * time.Second)
	defer timer.Stop()
	select {
	case <-done:
	case <-timer.C:
		_ = cmd.Process.Kill()
		<-done
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// WaitReady - read lines until {"ready":true}
// ─────────────────────────────────────────────────────────────────────────────

// WaitReady advances scanner until it finds a line that decodes as
// {"ready":true}.  Returns an error if the scanner ends before that.
func WaitReady(scanner *bufio.Scanner) error {
	for scanner.Scan() {
		var msg map[string]interface{}
		if err := json.Unmarshal(scanner.Bytes(), &msg); err != nil {
			continue // not valid JSON, keep reading
		}
		if ready, _ := msg["ready"].(bool); ready {
			return nil
		}
	}
	if err := scanner.Err(); err != nil {
		return fmt.Errorf("stitch: waiting for ready: %w", err)
	}
	return errors.New("stitch: child closed stdout before sending ready signal")
}

// ─────────────────────────────────────────────────────────────────────────────
// CallWithContext - context-aware RPC wait
// ─────────────────────────────────────────────────────────────────────────────

// CallWithContext sends a JSON-RPC request and waits for the response,
// honouring ctx cancellation. If ctx is cancelled before the response arrives,
// the pending entry is cleaned up and ctx.Err() is returned.
// The child sidecar continues executing the request - cancellation is Go-side only.
func CallWithContext(ctx context.Context, pending *PendingMap, id string, ch chan RpcResponse) (json.RawMessage, error) {
	select {
	case resp := <-ch:
		if resp.Error != nil {
			return nil, resp.Error
		}
		return resp.Result, nil
	case <-ctx.Done():
		pending.Delete(id)
		return nil, ctx.Err()
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// Ping - built-in health-check call
// ─────────────────────────────────────────────────────────────────────────────

// Ping sends a built-in __ping__ call and returns the raw response.
// It is a quick health-check for any bridge that supports the __ping__ method.
// The call uses a 5-second timeout via context.
func Ping(pending *PendingMap, write func(id, method string, params map[string]any) error) (json.RawMessage, error) {
	id := "ping-" + fmt.Sprintf("%d", time.Now().UnixNano())
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	ch := pending.Register(id)
	if err := write(id, "__ping__", nil); err != nil {
		pending.Delete(id)
		return nil, err
	}
	return CallWithContext(ctx, pending, id, ch)
}

// ─────────────────────────────────────────────────────────────────────────────
// BridgePool - least-connections pool of bridge instances
// ─────────────────────────────────────────────────────────────────────────────

// BridgePool routes calls across N bridge instances using least-connections.
// T is the bridge type (e.g. *PythonBridge). spawn is called N times at Start.
type BridgePool[T any] struct {
	workers  []T
	inFlight []int64
	mu       sync.Mutex
	size     int
	spawn    func() (T, error)
}

func NewBridgePool[T any](size int, spawn func() (T, error)) *BridgePool[T] {
	return &BridgePool[T]{size: size, spawn: spawn}
}

func (p *BridgePool[T]) Start() error {
	p.workers = make([]T, 0, p.size)
	p.inFlight = make([]int64, 0, p.size)
	for i := 0; i < p.size; i++ {
		w, err := p.spawn()
		if err != nil {
			return fmt.Errorf("pool start worker %d: %w", i, err)
		}
		p.workers = append(p.workers, w)
		p.inFlight = append(p.inFlight, 0)
	}
	return nil
}

// Pick returns the least-busy worker and its index, incrementing its in-flight counter.
func (p *BridgePool[T]) Pick() (T, int) {
	p.mu.Lock()
	defer p.mu.Unlock()
	minIdx := 0
	for i := 1; i < len(p.inFlight); i++ {
		if p.inFlight[i] < p.inFlight[minIdx] {
			minIdx = i
		}
	}
	p.inFlight[minIdx]++
	return p.workers[minIdx], minIdx
}

// Done decrements the in-flight counter for worker at idx.
func (p *BridgePool[T]) Done(idx int) {
	p.mu.Lock()
	p.inFlight[idx]--
	p.mu.Unlock()
}

File: shared/rust/bridge_client.rs
//! Stitch - shared Rust bridge client module.
//!
//! All Rust bridge clients (rust-python, rust-go, rust-ruby) include this
//! module (`mod bridge_client; use bridge_client::*;`) rather than duplicating
//! the pending-map, reader-thread spawn, kill helper, and error types.

use std::{
    collections::HashMap,
    io::{BufRead, BufReader},
    process::{Child, ChildStdout},
    sync::{
        mpsc::{sync_channel, SyncSender},
        Arc, Mutex,
    },
    thread::{self, JoinHandle},
};

use serde::{Deserialize, Serialize};
use serde_json::Value;

// ─────────────────────────────────────────────────────────────────────────────
// Wire types
// ─────────────────────────────────────────────────────────────────────────────

/// JSON-RPC error object carried in an error response from the sidecar.
#[derive(Debug, Deserialize, Clone)]
pub struct RpcError {
    pub code: Option<i64>,
    pub message: String,
    pub traceback: Option<String>,
    pub backtrace: Option<String>,
}

impl std::fmt::Display for RpcError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

/// A parsed response line from the sidecar (success or error, identified by id).
#[derive(Debug, Clone)]
pub struct RpcResponse {
    pub id: String,
    pub result: Option<Value>,
    pub error: Option<RpcError>,
}

// ─────────────────────────────────────────────────────────────────────────────
// PendingMap
// ─────────────────────────────────────────────────────────────────────────────

/// Shared pending-call map.  Both the caller thread and the reader thread hold
/// a clone of this `Arc` - the caller inserts, the reader removes and delivers.
pub type PendingMap = Arc<Mutex<HashMap<String, SyncSender<RpcResponse>>>>;

/// Create an empty PendingMap.
pub fn new_pending_map() -> PendingMap {
    Arc::new(Mutex::new(HashMap::new()))
}

// ─────────────────────────────────────────────────────────────────────────────
// Reader thread
// ─────────────────────────────────────────────────────────────────────────────

/// Spawn a daemon thread that reads newline-delimited JSON from `stdout`,
/// dispatches each response to the matching `SyncSender` in `pending`, and
/// signals `ready_tx` when the first `{"ready":true}` line is seen.
///
/// The returned `JoinHandle` should be stored in the bridge struct so the
/// thread is joined (or at least kept alive) for the bridge's lifetime.
pub fn spawn_reader_thread(
    stdout: ChildStdout,
    pending: PendingMap,
    ready_tx: SyncSender<()>,
) -> JoinHandle<()> {
    thread::spawn(move || {
        let reader = BufReader::new(stdout);
        let mut ready_sent = false;

        for line in reader.lines() {
            let raw = match line {
                Ok(l) => l,
                Err(e) => {
                    eprintln!("[bridge_client] reader IO error: {e}");
                    break;
                }
            };

            let trimmed = raw.trim();
            if trimmed.is_empty() {
                continue;
            }

            let v: Value = match serde_json::from_str(trimmed) {
                Ok(v) => v,
                Err(e) => {
                    eprintln!("[bridge_client] malformed JSON: {e} - `{trimmed}`");
                    continue;
                }
            };

            // Ready signal
            if !ready_sent && v.get("ready") == Some(&Value::Bool(true)) {
                ready_sent = true;
                let _ = ready_tx.send(());
                continue;
            }

            // Normal RPC response
            let id = match v.get("id").and_then(Value::as_str) {
                Some(s) => s.to_string(),
                None => {
                    eprintln!("[bridge_client] response missing `id`: {v}");
                    continue;
                }
            };

            let error = v.get("error").and_then(|e| {
                serde_json::from_value::<RpcError>(e.clone()).ok()
            });
            let result = v.get("result").cloned();
            let resp = RpcResponse { id: id.clone(), result, error };

            let mut map = pending.lock().unwrap();
            if let Some(tx) = map.remove(&id) {
                let _ = tx.send(resp);
            } else {
                eprintln!("[bridge_client] unknown response id: {id}");
            }
        }

        // Stdout closed - drain pending callers with an error.
        if !ready_sent {
            let _ = ready_tx.send(());
        }
        let mut map = pending.lock().unwrap();
        for (id, tx) in map.drain() {
            let _ = tx.send(RpcResponse {
                id,
                result: None,
                error: Some(RpcError {
                    code: Some(-32000),
                    message: "child process exited unexpectedly".to_string(),
                    traceback: None,
                    backtrace: None,
                }),
            });
        }
    })
}

// ─────────────────────────────────────────────────────────────────────────────
// kill_child helper
// ─────────────────────────────────────────────────────────────────────────────

/// Kill the child process and wait for it to exit.
/// On Unix: sends SIGKILL directly (callers should close stdin first for a
/// graceful shutdown before resorting to kill_child).
pub fn kill_child(child: &mut Child) {
    let _ = child.kill();
    let _ = child.wait();
}

// ─────────────────────────────────────────────────────────────────────────────
// Convenience: make a pending-slot and SyncSender pair
// ─────────────────────────────────────────────────────────────────────────────

/// Register a new call id in `pending` and return the receiver end of a
/// one-shot channel.  The reader thread will deliver to the sender end.
pub fn register_call(
    pending: &PendingMap,
    id: &str,
) -> std::sync::mpsc::Receiver<RpcResponse> {
    let (tx, rx) = sync_channel::<RpcResponse>(1);
    pending.lock().unwrap().insert(id.to_string(), tx);
    rx
}

// ─────────────────────────────────────────────────────────────────────────────
// ping_call convenience function
// ─────────────────────────────────────────────────────────────────────────────

/// Send a `__ping__` call and wait for `{"pong":true,"pid":<n>}`.
/// Returns `Ok(Value)` on success or `Err(String)` on timeout / error.
/// `write_fn` sends a serialised request line to the child's stdin.
pub fn ping_call(
    pending: &PendingMap,
    write_fn: impl Fn(&str) -> Result<(), String>,
) -> Result<Value, String> {
    use std::time::Duration;

    let id = format!("ping-{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .subsec_nanos());

    let rx = register_call(pending, &id);

    let req = serde_json::json!({
        "id": &id,
        "method": "__ping__",
        "params": {}
    });
    write_fn(&req.to_string())?;

    let resp = rx
        .recv_timeout(Duration::from_secs(5))
        .map_err(|_| "ping timed out".to_string())?;

    if let Some(err) = resp.error {
        return Err(format!("ping error: {}", err.message));
    }

    resp.result.ok_or_else(|| "ping: no result".to_string())
}

// ─────────────────────────────────────────────────────────────────────────────
// Streaming support
// ─────────────────────────────────────────────────────────────────────────────

/// A single frame from a streaming sidecar response.
#[derive(Debug)]
pub struct StreamFrame {
    /// The chunk data, present for intermediate frames.
    pub chunk: Option<serde_json::Value>,
    /// Set on the terminal frame (empty Value = stream done).
    pub result: Option<serde_json::Value>,
    /// Set if the stream ended with an error.
    pub error: Option<RpcError>,
}

/// A streaming response iterator yielded by `stream_call`.
/// Each `next()` blocks until the next chunk or terminal frame arrives.
pub struct BridgeStream {
    rx: std::sync::mpsc::Receiver<StreamFrame>,
}

impl BridgeStream {
    pub fn new(rx: std::sync::mpsc::Receiver<StreamFrame>) -> Self {
        Self { rx }
    }
}

impl Iterator for BridgeStream {
    type Item = Result<serde_json::Value, RpcError>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.rx.recv() {
            Ok(frame) => {
                if let Some(err) = frame.error {
                    return Some(Err(err));
                }
                if frame.result.is_some() {
                    return None; // terminal frame = end of stream
                }
                frame.chunk.map(Ok)
            }
            Err(_) => None,
        }
    }
}

/// Shared pending-stream map. Keyed by call id; each entry is a bounded sender
/// for `StreamFrame` values delivered by `spawn_streaming_reader_thread`.
pub type StreamPendingMap = Arc<Mutex<HashMap<String, SyncSender<StreamFrame>>>>;

/// Create an empty StreamPendingMap.
pub fn new_stream_pending_map() -> StreamPendingMap {
    Arc::new(Mutex::new(HashMap::new()))
}

/// Register a new streaming call id in `stream_pending` and return the
/// receiver end of a bounded channel.  The streaming reader thread will
/// deliver `StreamFrame`s to the sender end.
pub fn register_stream_call(
    stream_pending: &StreamPendingMap,
    id: &str,
) -> std::sync::mpsc::Receiver<StreamFrame> {
    let (tx, rx) = std::sync::mpsc::sync_channel::<StreamFrame>(64);
    stream_pending.lock().unwrap().insert(id.to_string(), tx);
    rx
}

/// Like `spawn_reader_thread` but handles both regular and streaming calls.
///
/// - Lines with a `"chunk"` key (and no `"result"` or `"error"`) are chunk
///   frames and are delivered to `stream_pending`.
/// - Lines whose id is present in `stream_pending` (terminal / error on
///   stream) are delivered there and the entry removed.
/// - All other lines are delivered to the regular `pending` map.
pub fn spawn_streaming_reader_thread(
    stdout: ChildStdout,
    pending: PendingMap,
    stream_pending: StreamPendingMap,
    ready_tx: SyncSender<()>,
) -> JoinHandle<()> {
    thread::spawn(move || {
        let reader = BufReader::new(stdout);
        let mut ready_sent = false;

        for line in reader.lines() {
            let raw = match line {
                Ok(l) => l,
                Err(_) => break,
            };
            let trimmed = raw.trim();
            if trimmed.is_empty() {
                continue;
            }

            let v: Value = match serde_json::from_str(trimmed) {
                Ok(v) => v,
                Err(e) => {
                    eprintln!("[bridge_client] malformed JSON: {e}");
                    continue;
                }
            };

            // Ready signal
            if !ready_sent && v.get("ready") == Some(&Value::Bool(true)) {
                ready_sent = true;
                let _ = ready_tx.send(());
                continue;
            }

            let id = match v.get("id").and_then(Value::as_str) {
                Some(s) => s.to_string(),
                None => continue,
            };

            // Chunk frame (has "chunk", no "result" / "error")
            if v.get("chunk").is_some()
                && v.get("result").is_none()
                && v.get("error").is_none()
            {
                let sp = stream_pending.lock().unwrap();
                if let Some(tx) = sp.get(&id) {
                    let frame = StreamFrame {
                        chunk: v.get("chunk").cloned(),
                        result: None,
                        error: None,
                    };
                    let _ = tx.send(frame);
                }
                continue;
            }

            let error = v
                .get("error")
                .and_then(|e| serde_json::from_value::<RpcError>(e.clone()).ok());
            let result = v.get("result").cloned();

            // Terminal stream frame or stream error
            {
                let mut sp = stream_pending.lock().unwrap();
                if let Some(tx) = sp.remove(&id) {
                    let frame = StreamFrame { chunk: None, result, error };
                    let _ = tx.send(frame);
                    continue;
                }
            }

            // Regular pending call
            let resp = RpcResponse { id: id.clone(), result, error };
            let mut map = pending.lock().unwrap();
            if let Some(tx) = map.remove(&id) {
                let _ = tx.send(resp);
            } else {
                eprintln!("[bridge_client] unknown response id: {id}");
            }
        }

        // Drain remaining callers with a "child exited" error.
        if !ready_sent {
            let _ = ready_tx.send(());
        }
        let mut map = pending.lock().unwrap();
        for (id, tx) in map.drain() {
            let _ = tx.send(RpcResponse {
                id,
                result: None,
                error: Some(RpcError {
                    code: Some(-32000),
                    message: "child exited".to_string(),
                    traceback: None,
                    backtrace: None,
                }),
            });
        }
        let mut sp = stream_pending.lock().unwrap();
        for (_, tx) in sp.drain() {
            let _ = tx.send(StreamFrame {
                chunk: None,
                result: None,
                error: Some(RpcError {
                    code: Some(-32000),
                    message: "child exited".to_string(),
                    traceback: None,
                    backtrace: None,
                }),
            });
        }
    })
}

// ─────────────────────────────────────────────────────────────────────────────
// SupervisedBridge wrapper
// ─────────────────────────────────────────────────────────────────────────────

/// Wraps a bridge and restarts the child on unexpected exit.
///
/// # Example
/// ```rust,no_run
/// let supervised = SupervisedBridge::new(|| GoBridge::spawn("./sidecar", &[]), 3);
/// ```
pub struct SupervisedBridge<T, F>
where
    F: Fn() -> Result<T, String> + Send + 'static,
{
    factory: F,
    max_restarts: u32,
    restart_count: u32,
    inner: Option<T>,
}

impl<T, F> SupervisedBridge<T, F>
where
    F: Fn() -> Result<T, String> + Send + 'static,
{
    pub fn new(factory: F, max_restarts: u32) -> Result<Self, String> {
        let inner = factory()?;
        Ok(Self {
            factory,
            max_restarts,
            restart_count: 0,
            inner: Some(inner),
        })
    }

    /// Attempt to restart the inner bridge after unexpected failure.
    /// Returns Err if max_restarts has been reached.
    pub fn restart(&mut self) -> Result<(), String> {
        if self.restart_count >= self.max_restarts {
            return Err(format!(
                "max restarts ({}) reached",
                self.max_restarts
            ));
        }
        let delay_ms = 100u64 * (1 << self.restart_count.min(7));
        std::thread::sleep(std::time::Duration::from_millis(delay_ms));
        self.restart_count += 1;
        eprintln!(
            "[supervised] restarting child (attempt {}/{})",
            self.restart_count, self.max_restarts
        );
        self.inner = Some((self.factory)()?);
        Ok(())
    }

    /// Get a reference to the inner bridge.
    pub fn inner(&self) -> Option<&T> {
        self.inner.as_ref()
    }

    /// Get a mutable reference to the inner bridge.
    pub fn inner_mut(&mut self) -> Option<&mut T> {
        self.inner.as_mut()
    }
}
